### PR TITLE
Feature/96 popularity prior

### PR DESCRIPTION
**Pull request by rcffc**
Depends on #602.
For each candidate, look how often it's label appears in the text.
Note that this is different from what was original proposed, namely getting the count of  actual linkings in the text.

### CODE REVIEW

File: inception-concept-linking/pom.xml [hunk 1]
```diff
@@ -86,8 +86,13 @@
       <groupId>de.tudarmstadt.ukp.clarin.webanno</groupId>
       <artifactId>webanno-api</artifactId>
     </dependency>
-
     <dependency>
+      <groupId>de.tudarmstadt.ukp.clarin.webanno</groupId>
+      <artifactId>webanno-security</artifactId>
+    </dependency>
+
```
Comment (Rentier) [inline inception-concept-linking/pom.xml]: The whitespace looks wrong here (indent and blank lines)

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/model/CandidateEntity.java [hunk 2]
```diff
@@ -84,6 +84,11 @@
      */
     private double idRank;
 
+    /**
+     *
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/model/CandidateEntity.java]: There is no comment here

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 3]
```diff
@@ -146,7 +151,18 @@ public void afterPropertiesSet() throws Exception
...
     }
-    
+
+    private List<SearchResult> getSearchResults(User aUser, Project aProject, String aQuery)
+    {
+        try {
+            return searchService.query(aUser, aProject, aQuery);
+        }
+        catch (Exception e) {
+            log.debug("Error in the query.", e);
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: Isnt this better an error?

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 4]
```diff
@@ -397,22 +426,25 @@ private synchronized Sentence getMentionSentence(JCas aJcas, int aBegin)
...
-                        }
+
+        if (aKB.getType() == RepositoryType.REMOTE && entityFrequencyMap != null) {
+            RepositoryImplConfig cfg = kbService.getKnowledgeBaseConfig(aKB);
+            if (((SPARQLRepositoryConfig) cfg).getQueryEndpointUrl()
+                .equals(IriConstants.UKP_WIKIDATA_SPARQL_ENDPOINT)) {    // For UKP Wikidata
+                for (CandidateEntity l : aCandidatesFullText) {
+                    String key = l.getIRI();
+                    key = key.replace(IriConstants.PREFIX_WIKIDATA_ENTITY, "");
+                    if (entityFrequencyMap.get(key) != null) {
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: Why not `Map::containsKey`?

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 5]
```diff
@@ -397,22 +426,25 @@ private synchronized Sentence getMentionSentence(JCas aJcas, int aBegin)
...
-                if (aKB.getType() == RepositoryType.REMOTE
-                    && aKB.getFullTextSearchIri().equals(IriConstants.FTS_VIRTUOSO)) {
-                    RepositoryImplConfig cfg = kbService.getKnowledgeBaseConfig(aKB);
-                    if (((SPARQLRepositoryConfig) cfg).getQueryEndpointUrl()
-                        .equals(IriConstants.UKP_WIKIDATA_SPARQL_ENDPOINT)) {
-                        key = key.replace(IriConstants.PREFIX_WIKIDATA_ENTITY, "");
-                        if (entityFrequencyMap.get(key) != null) {
-                            l.setFrequency(entityFrequencyMap.get(key));
-                        }
+
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: Could you comment on what happens here?

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 6]
```diff
@@ -397,22 +426,25 @@ private synchronized Sentence getMentionSentence(JCas aJcas, int aBegin)
...
+                for (CandidateEntity l : aCandidatesFullText) {
+                    String key = l.getIRI();
+                    key = key.replace(IriConstants.PREFIX_WIKIDATA_ENTITY, "");
+                    if (entityFrequencyMap.get(key) != null) {
+                        l.setFrequency(entityFrequencyMap.get(key));
                     }
                 }
             }
+        } else if (aKB.getType() == RepositoryType.LOCAL) {
+            for (CandidateEntity l : aCandidatesFullText) {
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: I would put that into a function called `calculateLabelFrequencyPopularityPrior` or so

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java [hunk 7]
```diff
@@ -397,22 +426,25 @@ private synchronized Sentence getMentionSentence(JCas aJcas, int aBegin)
...
+                List<SearchResult> results 
+                    = getSearchResults(aUser, aKB.getProject(), l.getAltLabel());
+                l.setIdRank(results.size());
```
Comment (Rentier) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/ConceptLinkingService.java]: I would rename `IdRank` to Popularity or so, as we now have more priors than Wiki id

**Comment by Rentier** [comment]
Did you branch this from master? I ask, because it depends on the bugfix for concepts without labels.

**Comment by reckart** [comment]
> For each candidate, look how often it's label appears in the text.

The full text search allows us to query for how often a particular candidate has been linked, e.g. using `<Named_entity.identifier-exact="candidate-IRI"/>` - and such a query should be used. Just counting how often the label appears in the text IMHO makes no sense.

**Comment by reckart** [comment]
@jcklie I guess we can close this PR and create a new one based on the now-(more-or-less)-modularized entity ranking architecture.

**Pull request closed without merging**